File: Semester3/Advanced Methods of Programming/Toy Language Interpreter C#/Assignement7/Assignement7/Model/ArithmeticExpression.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Assignement7.Collection;

namespace Assignement7.Model
{
    public class ArithmeticExpression : Expression
    {
        public enum Operation
        {
            ADD,
            SUBTRACT,
            MULTIPLY,
            DIVIDE
        }

        private Expression left;
        private Expression right;
        Operation op;
        
        public ArithmeticExpression(Expression left, Expression right, Operation op)
        {
            this.op = op;
            this.left = left;
            this.right = right;
        }

        public override int Evaluate(MyIDictionary<string, int> symbolTable)
        {
            int l = this.left.Evaluate(symbolTable);
            int r = this.right.Evaluate(symbolTable);

            int ret = l;
            switch(this.op)
            {
                case Operation.ADD:
                    ret += r;
                    break;
                case Operation.SUBTRACT:
                    ret -= r;
                    break;
                case Operation.MULTIPLY:
                    ret *= r;
                    break;
                case Operation.DIVIDE:
                    if (r == 0)
                        throw new Exception("Division by zero");
                    ret /= r;
                    break;
            }
            return ret;
        }

        public override string ToString()
        {
            string ret = this.left.ToString();

            switch (this.op)
            {
                case Operation.ADD:
                    ret += " + ";
                    break;
                case Operation.SUBTRACT:
                    ret += " - ";
                    break;
                case Operation.MULTIPLY:
                    ret += " * ";
                    break;
                case Operation.DIVIDE:
                    ret += " / ";
                    break;
            }

            ret += this.right.ToString();
            return ret;
        }
    }
}

File: Semester3/Advanced Methods of Programming/Toy Language Interpreter JavaFx/Assignment8/src/Model/Expression/ArithmeticExpression.java
package Model.Expression;

import Collection.Dictionary.MyIDictionary;
import Model.Exceptions.*;

public class ArithmeticExpression extends Expression {
    private Expression exp1;
    private Expression exp2;
    private OperationEnum operation;

    public ArithmeticExpression(Expression exp1, Expression exp2, OperationEnum operation){
        this.exp1 = exp1;
        this.exp2 = exp2;
        this.operation = operation;
    }

    @Override
    public Integer evaluate(MyIDictionary<String, Integer> symbolTable, MyIDictionary<Integer, Integer> heapTable) throws ADTEmptyException, DivisionByZero, VariableNotFoundException, ComparisonExpression {
        Integer eval1 = exp1.evaluate(symbolTable, heapTable);
        Integer eval2 = exp2.evaluate(symbolTable, heapTable);

        if(operation == OperationEnum.PLUS) return eval1 + eval2;
        else if(operation == OperationEnum.MINUS) return eval1 - eval2;
        else if(operation == OperationEnum.MULTIPLY) return eval1 * eval2;
        else if(operation == OperationEnum.DIVIDE){
            if(eval2 == 0) throw new DivisionByZero("Error trying to divide by zero.");
            return eval1 / eval2;
        }
        return -1;
    }

    @Override
    public String toString(){
        String result = "( " + exp1.toString();
        if(operation == OperationEnum.PLUS) result += " + ";
        else if(operation == OperationEnum.MINUS) result += " - ";
        else if(operation == OperationEnum.MULTIPLY) result += " * ";
        else if(operation == OperationEnum.DIVIDE) result += " / ";
        result += exp2.toString() + " )";
        return result;
    }
}
